File: code/go/02_linked_list/linked_list/linked_list.go
package linked_list

import "fmt"

type ListNode struct {
	Val  int
	Next *ListNode
}

type LinkedList struct {
	Head *ListNode
	Size int
}

func Constructor() *LinkedList {
	return &LinkedList{
		Head: &ListNode{Val: 0, Next: nil},
		Size: 0,
	}
}

// Get 获取链表中第 index 个节点的值。如果索引无效，则返回-1
func (l *LinkedList) Get(index int) int {
	if index < 0 || index >= l.Size {
		return -1
	}

	cur := l.Head
	for index >= 0 {
		cur = cur.Next
		index--
	}

	return cur.Val
}

// AddAtHead 在链表的第一个元素之前添加一个值为 val 的节点。插入后，新节点将成为链表的第一个节点。
func (l *LinkedList) AddAtHead(val int) {
	node := &ListNode{Val: val}
	node.Next = l.Head.Next
	l.Head.Next = node
	l.Size++
}

// AddAtTail 将值为 val 的节点追加到链表的最后一个元素
func (l *LinkedList) AddAtTail(val int) {
	node := &ListNode{Val: val}
	current := l.Head
	for current.Next != nil {
		current = current.Next
	}
	current.Next = node
	l.Size++
}

// AddAtIndex 在链表中的第 index 个节点之前添加值为 val  的节点。
// 如果 index 等于链表的长度，则该节点将附加到链表的末尾。
// 如果 index 大于链表长度，则不会插入节点。如果index小于0，则在头部插入节点。
func (l *LinkedList) AddAtIndex(index int, val int) {
	if index > l.Size {
		return
	}

	if index < 0 {
		index = 0
	}

	current := l.Head
	for index > 0 {
		current = current.Next
		index--
	}

	node := &ListNode{Val: val}
	node.Next = current.Next
	current.Next = node
	l.Size++
}

// DeleteAtIndex 如果索引 index 有效，则删除链表中的第 index 个节点
func (l *LinkedList) DeleteAtIndex(index int) {
	if index < 0 || index >= l.Size {
		return
	}

	slow := l.Head
	fast := l.Head.Next

	for index > 0 {
		fast = fast.Next
		slow = slow.Next
		index--
	}
	slow.Next = fast.Next
	l.Size--
}

func (l *LinkedList) Print() {
	current := l.Head
	for current.Next != nil {
		fmt.Println(current.Next.Val)
		current = current.Next
	}
}

File: src/code/python/02_linked_list/02_linked_list/linked_list.py
import unittest


class ListNode:
    """ "list node"""

    def __init__(self, val: int = 0, next=None) -> None:
        self.val = val
        self.next = next


class LinkedList:
    """
    设计链表 https://leetcode.cn/problems/design-linked-list/
    """

    def __init__(self) -> None:
        self.head = ListNode(0)
        self.size = 0

    def get(self, index: int) -> int:
        """获取链表中第 index 个节点的值。如果索引无效，则返回-1"""
        if index < 0 or index >= self.size:
            return -1

        current = self.head.next
        if current == None:
            return -1

        while index > 0:
            current = current.next
            index -= 1
        return current.val

    def add_at_head(self, val: int):
        """在链表的第一个元素之前添加一个值为 val 的节点。插入后，新节点将成为链表的第一个节点。"""
        node = ListNode(val=val)
        node.next = self.head.next
        self.head.next = node
        self.size += 1

    def add_at_tail(self, val: int):
        """将值为 val 的节点追加到链表的最后一个元素"""
        node = ListNode(val=val)
        current = self.head
        while current.next != None:
            current = current.next
        current.next = node
        self.size += 1

    def add_at_index(self, index: int, val: int):
        """
        在链表中的第 index 个节点之前添加值为 val  的节点。
        如果 index 等于链表的长度，则该节点将附加到链表的末尾。
        如果 index 大于链表长度，则不会插入节点。如果index小于0，则在头部插入节点
        """
        if index < 0 or index > self.size:
            return

        current = self.head
        while index > 0:
            current = current.next
            index -= 1

        node = ListNode(val=val, next=current.next)
        current.next = node
        self.size += 1

    def delete_at_index(self, index: int):
        """如果索引 index 有效，则删除链表中的第 index 个节点"""
        if index < 0 or index >= self.size:
            return

        slow = self.head
        fast = self.head.next

        while index > 0:
            slow = slow.next
            fast = fast.next
            index -= 1

        slow.next = fast.next
        self.size -= 1

    def print(self):
        current = self.head
        while current.next != None:
            print(current.next.val)
            current = current.next


class TestLinkedList(unittest.TestCase):
    def test_linked_list(self):
        linked_list = LinkedList()
        linked_list.add_at_head(1)
        linked_list.add_at_tail(3)
        linked_list.add_at_index(1, 2)
        linked_list.add_at_index(2, 4)
        linked_list.delete_at_index(3)
        linked_list.print()
        print(linked_list.get(1))


if __name__ == "__main__":
    unittest.main()
